### PR TITLE
New TypeCast

### PR DESCRIPTION
**Pull request by rmorshea**
Refactoring #317

### CODE REVIEW

**Comment by rmorshea** [comment]
@minrk if you have any thoughts on why the "hard part" (a0296cf) is failing let me know.

Something subtle is wrong, so I'm tempted to just submit the "easy part" (6b66f1c) on it's own.

**Comment by rmorshea** [comment]
Ugh... I just forgot to pass the default value through `super().__init__` :expressionless: - not nearly as subtle an error as I thought. Still have to make `Dict` inherit from `TypeCast` though.

**Comment by rmorshea** [comment]
@minrk passing now.

**Comment by rmorshea** [comment]
@minrk questions, comments, or concerns about this? I mostly need to add documentation.

**Comment by minrk** [comment]
Sorry I haven't gotten to it, yet. This is a big rewrite of a lot of traitlets, which always makes me nervous. I just haven't had the time to go through it thoroughly and test it with every downstream IPython and Jupyter project, etc.. I'll try to get to this soon.

**Comment by rmorshea** [comment]
@minrk, the traits of most concern would be those inheriting from 'Container' since their implementation is relatively complicated.

**Comment by rmorshea** [comment]
@minrk, would it be better for me to break this into multiple PRs?

1. Introduce `TypeCast` without refactoring
2. Refactor trivial cases (e.g. `Int`, `Unicode`, etc.)
3. Refactor hard cases (e.g. `Container`, `Mapping`, etc.)

**Comment by rmorshea** [comment]
@minrk 6.0? or 5.0?

File: traitlets/traitlets.py [hunk 1]
```diff
@@ -2269,95 +2291,110 @@ def validate(self, obj, value):
...
+    def __init__(self, *args, **kwargs):
+        super(Mutable, self).__init__(*args, **kwargs)
+        self._default_progenitor = self.default_value
+        self.default_value = Undefined
 
-        trait : TraitType [ optional ]
-            the type for restricting the contents of the Container.  If unspecified,
-            types are not checked.
+    def make_dynamic_default(self):
+        return copy.copy(self._default_progenitor)
```
Comment (minrk) [inline traitlets/traitlets.py]: Shouldn't this be `copy.copy(self.default_value)`? Why is the default_value unused and a separate private attribute used for the same purpose?

File: traitlets/traitlets.py [hunk 2]
```diff
@@ -2269,95 +2291,110 @@ def validate(self, obj, value):
...
-class Container(Instance):
-    """An instance of a container (list, set, etc.)
-
-    To be subclassed by overriding klass.
-    """
-    klass = None
-    _cast_types = ()
-    _valid_defaults = SequenceTypes
-    _trait = None
+class Mutable(TypeCast):
```
Comment (minrk) [inline traitlets/traitlets.py]: What is Mutable? If it's only here to be a base class, make it private: `_Mutable`. Also make sure that every one of these has a docstring.

File: traitlets/traitlets.py [hunk 3]
```diff
@@ -2269,95 +2291,110 @@ def validate(self, obj, value):
...
             except TraitError as error:
                 self.error(obj, v, error)
             else:
-                validated.append(v)
-        return self.klass(validated)
+                new.append(v)
+        return new
+
+
+class BaseSequence(Container):
```
Comment (minrk) [inline traitlets/traitlets.py]: What is BaseSequence? It needs a docstring, and should be private if it's only here to be a helper for Sequence below.

File: traitlets/tests/test_traitlets.py [hunk 4]
```diff
@@ -1585,8 +1585,8 @@ def coerce(self, value):
         return value
 
     def test_invalid_args(self):
-        self.assertRaises(TypeError, Tuple, 5)
```
Comment (minrk) [inline traitlets/tests/test_traitlets.py]: Calling a constructor with the wrong arguments does normally raise TypeError. Why has this changed?

**Comment by rmorshea** [comment]
Close in favor of https://github.com/ipython/traitlets/pull/477

**Pull request closed without merging**